dali/pipeline/operators/reader/parser/coco_parser.h

```diff
@@ -0,0 +1,97 @@
+// Copyright (c) 2017-2018, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_PIPELINE_OPERATORS_READER_PARSER_COCO_PARSER_H_
+#define DALI_PIPELINE_OPERATORS_READER_PARSER_COCO_PARSER_H_
+
+#include <map>
+#include <string>
+#include <vector>
+
+#include "dali/pipeline/operators/reader/parser/parser.h"
+
+namespace dali {
+
+struct Annotation {
+  /*
+   * Currently, only bboxes and labels are supported,
+   * while key points and segmentation masks need to be handled
+   * somehow in the future
+   */
+  Annotation(float x, float y, float w, float h, int category_id)
+    : bbox({x, y, w, h}),
+      category_id(category_id) {}
+
+  // bbox contains 5 values: x, y, W, H, category_id
+  std::vector<float> bbox;
+  int category_id;
+  friend std::ostream& operator<<(std::ostream& os, Annotation& an);
+};
+
+std::ostream& operator<<(std::ostream& os, Annotation& an) {
+  std::vector<float>& bbox = an.bbox;
+  os << "Annotation(category_id=" << an.category_id
+  << ",bbox = [" << bbox[0] << "," << bbox[1]
+  << "," << bbox[2] << "," << bbox[3] << "])";
+  return os;
+}
+
+using AnnotationMap = std::multimap<int, Annotation>;
+
+class COCOParser: public Parser {
+ public:
+  explicit COCOParser(const OpSpec& spec, const AnnotationMap& annotations_multimap)
+    : Parser(spec),
+    annotations_multimap_(annotations_multimap) {}
+
+  void Parse(const uint8_t* data, const size_t size, SampleWorkspace* ws) override {
+    Index image_size = size - sizeof(int);
+    auto *image_output = ws->Output<CPUBackend>(0);
+    auto *bbox_output = ws->Output<CPUBackend>(1);
+    auto *label_output = ws->Output<CPUBackend>(2);
+    int image_id =
+         *reinterpret_cast<const int*>(data + image_size);
```

Comment:
Sadly, this is UB, and should be done in similar fashion as: https://chromium.googlesource.com/chromium/src/+/1587f7d/base/macros.h#76

But still, it will work on all compilers.

Reply:
Right.